### PR TITLE
Prox approx part 3

### PR DESCRIPTION
**Pull request by bknueven**
- Re-implement cut manager based on tolerance
- Add cuts based on W
- Add safety cuts if W == 0 and xbar == xhat
- Don't bother if the variable is fixed
- Skip cut if it is outside the variable bounds

### CODE REVIEW

File: mpisppy/utils/sputils.py [hunk 1]
```diff
@@ -984,7 +984,8 @@ def nonant_cost_coeffs(s):
         if id(var) in s._mpisppy_data.varid_to_nonant_index:
             raise RuntimeError(
                 "Found nonlinear variables in the objective function. "
```
Comment (DLWoodruff) [inline mpisppy/utils/sputils.py:986]: Change "Found nonlinear..." to "A call to nonant_cost_coefficient found nonlinear" (even though the user could figure that out from the stack trace, I don't want to give the impression that nonlinear is not OK in general")

**Review by DLWoodruff — APPROVED** [review]

**Comment by bknueven** [comment]
Not really sure if this PR or #467 should be merged instead.

PR #467 takes a more systematic approach, and hence adds more cuts. This can slow subproblem solves, but the iterations seem to be of higher quality due to the extra cuts.

**Comment by DLWoodruff** [comment]
> Not really sure if this PR or #467 should be merged instead.
> 
> PR #467 takes a more systematic approach, and hence adds more cuts. This can slow subproblem solves, but the iterations seem to be of higher quality due to the extra cuts.

To state the obvious: we could have both and that might even be a good idea (but does require even more work)

File: mpisppy/utils/sputils.py (hunk 1 above)
Comment (bknueven) [inline mpisppy/utils/sputils.py:986]:
```suggestion
                "A call to nonant_cost_coefficient found nonlinear variables in the objective function. "
```

**Comment by bknueven** [comment]
> > Not really sure if this PR or #467 should be merged instead.
> > PR #467 takes a more systematic approach, and hence adds more cuts. This can slow subproblem solves, but the iterations seem to be of higher quality due to the extra cuts.
> 
> To state the obvious: we could have both and that might even be a good idea (but does require even more work)

Yes. I am confident this is an improvement over `main`, so let's merge this. It is simpler, and then we can consider the additional changes in #467 separately.

**Pull request merged**